sonar-core/src/main/java/org/sonar/core/issue/tracking/Tracker.java

```diff
@@ -0,0 +1,299 @@
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.core.issue.tracking;
+
+import com.google.common.base.Predicate;
+import com.google.common.base.Strings;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.rule.RuleKey;
+
+import static com.google.common.collect.FluentIterable.from;
+
+public class Tracker<RAW extends Trackable, BASE extends Trackable> {
+
+  public Tracking<RAW, BASE> track(Input<RAW> rawInput, Input<BASE> baseInput) {
+    Tracking<RAW, BASE> tracking = new Tracking<>(rawInput, baseInput);
+
+    // 1. match issues with same rule, same line and same line hash, but not necessarily with same message
+    match(tracking, LineAndLineHashKeyFactory.INSTANCE);
+
+    // 2. detect code moves by comparing blocks of codes
+    detectCodeMoves(rawInput, baseInput, tracking);
+
+    // 3. match issues with same rule, same message and same line hash
+    match(tracking, LineHashAndMessageKeyFactory.INSTANCE);
+
+    // 4. match issues with same rule, same line and same message
+    match(tracking, LineAndMessageKeyFactory.INSTANCE);
+
+    // 5. match issues with same rule and same line hash but different line and different message.
+    // See SONAR-2812
+    match(tracking, LineHashKeyFactory.INSTANCE);
+
+    // TODO what about issues on line 0 ?
+    relocateManualIssues(rawInput, tracking);
+
+    return tracking;
+  }
+
+  private void detectCodeMoves(Input<RAW> rawInput, Input<BASE> baseInput, Tracking<RAW, BASE> tracking) {
+    if (!tracking.isComplete()) {
+      new BlockRecognizer<RAW, BASE>().match(rawInput, baseInput, tracking);
+    }
+  }
+
+  private void match(Tracking<RAW, BASE> tracking, SearchKeyFactory factory) {
+    if (tracking.isComplete()) {
+      return;
+    }
+
+    Multimap<SearchKey, BASE> baseSearch = ArrayListMultimap.create();
+    for (BASE base : tracking.getUnmatchedBases()) {
+      baseSearch.put(factory.create(base), base);
+    }
+
+    Collection<RAW> trackedRaws = new ArrayList<>();
+    for (RAW raw : tracking.getUnmatchedRaws()) {
+      SearchKey rawKey = factory.create(raw);
+      Collection<BASE> bases = baseSearch.get(rawKey);
+      if (!bases.isEmpty()) {
+        // TODO taking the first one. Could be improved if there are more than 2 issues on the same line.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)